Add usage documentation for Rich Text package (#35885)
WordPress/gutenberg: packages/rich-text/README.md

_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for such language features and APIs, you should include [the polyfill shipped in `@wordpress/babel-preset-default`](https://github.com/WordPress/gutenberg/tree/HEAD/packages/babel-preset-default#polyfill) in your code._

## Usage

> The Rich Text package is designed to facilitate the application and manipulation of rich formatting onto strings of plain text (e.g. bold, italic, links). This is facilitated via the usage of a `RichTextValue` value object (referred to from here on as `value`) which, by separating text from formatting, affords the ability to easily search and manipulate rich formats.

Reviewer: ```suggestion
The Rich Text package applies rich formatting onto plain text strings (e.g. bold, italic, links). By using a `RichTextValue` value object (referred to from here on as `value`) separates the text from formatting, making it easier to search and manipulate rich formats.
```
I edited the intro, reworded here and there for clarity. I think still conveys the same info, let me know.

Author: I agree this was overly "wordy" 😆 

I want to convey that the package _aids_ in applying rich formatting. I'll tweak you're great suggested improvement.

Author: Amended in https://github.com/WordPress/gutenberg/pull/35885/commits/c66a29f8f9230c9c98f19de2abf7da4fc0f2af62

### The RichTextValue object

The value object is comprised of the following:

-   `text` - the string of text to which rich formats are to be applied.
-   `formats` - a sparse array of the same length as `text` that is filled with [formats](https://developer.wordpress.org/block-editor/how-to-guides/format-api/) (e.g. `core/link`, `core/bold` etc.) at the positions where the text is formatted.
-   `start` - an index in the `text` representing the _start_ of the currently active selection.
-   `end` - an index in the `text` representing the _end_ of the currently active selection.

> You should not attempted to create your own `value` objects. Rather you should rely on the built in methods of the `@wordpress/rich-text` package to build these for you.

Reviewer: ```suggestion
You should not attempt to create your own `value` objects. Rather you should rely on the built in methods of the `@wordpress/rich-text` package to build these for you.
```

> Understanding how a value represents richly formatted text is very important so let's consider an example.

Reviewer: ```suggestion
It is important to understand how a value represents richly formatted text. Here is an example to illustrate.
```

> If `text` is formatted from position 2-5 in bold (`core/bold`) and from position 2-10 with a link (`core/link`), then you'll find:

Reviewer: ```suggestion
If `text` is formatted from position 2-5 in bold (`core/bold`) and from position 2-8 with a link (`core/link`), then you'll find:
```

I think you use 8 in your example below.

      {
        type: 'core/bold',
      },
      {
        type: 'core/link',
      }
    ],
    [
      {
        type: 'core/bold',
      },
      {
        type: 'core/link',
      }
    ],
    [
      {
        type: 'core/bold',
      },
      {
        type: 'core/link',
      }
    ],
    [ // 6
      {
        type: 'core/link',
      }
    ]
    [
      {
        type: 'core/link',
      }
    ],
    [
      {
        type: 'core/link',
      }
    ],
    [], // 9
    [], // 10
    [], // 11
  ]
}
```

### Selections

Let's continue to consider the above example with the text `Hello world`.

If, as a user, I make a selection of the word `Hello` this would result in a value object with `start` and `end` as `0` and `5` respectively.

In general, this is useful for knowing which portion of the text is selected. However, we need to consider that selections may also be "collapsed".

#### Collapsed selections

A collapsed selection is one where `start` and `end` values are identical. This happens when no characters are selected, but there is a caret present. This most often occurs when a user places the cursor/caret within a string of text but does not make a selected range.

> Given that it has no range identifiers, finding the currently selected portion of text from collapsed values can be challenging.

Author: @mkaz I think this is what confused you. I agree it's completely unclear.

What I'm trying to convey is that as a collapsed selection has `start` and `end` that are _identical_ there is no "range" between them (unlike say, if you had `start: 1` and `end: 5`). The problem with this is that `start` and `end` are useful for identifying what portions of the `text` from the `RichTextValue` are _selected_ by the user. Unfortunately collapsed selections mean we loose this important info.

I believe Ella is looking to address this via modifications but it currently represents a challenge.

Author: Addressed simplification of wording in https://github.com/WordPress/gutenberg/pull/35885/commits/28412590dbbcc458887e9920a9dffe4d38121a4e

### Replacements

> TODO: What are replacements? How do they work?

Author: @gwwar Do you understand the concept of replacements? I feel like you do. Might you be able to help me fill in the blanks here?

Author: I've removed this section. Shall we tackle in a follow up?
